### ISSUE TITLE
NP architecture is described differently in "Attentive Neural Processes" paper

### ISSUE THREAD

**Issue by christabella** [EmilienDupont/neural-processes#2]
Thank you for the wonderful implementation! I have a high-level question:

The code looks like it is following the latent variable formulation of Conditional Neural Processes (CNP), which also sounds exactly like the description in section 2.4 of the NP paper:
![latent CNP](https://bella.cc/ox-hugo/screenshot_20190516_234937.png)
* Encoding context pairs `(x, y)` as representation `r` in `xy_to_r`
* Transforming `r` to latent parameters of `z` Normal in `r_to_mu_sigma`
* Decoding pairs `(x, z)` to parameters of `y` Normal in `xz_to_y`

However, the NP model described in the "Attentive Neural Processes" (ANP) paper has a few differences:
![NP](https://bella.cc/ox-hugo/screenshot_20190517_012125.png)
1. The `Encoder` (`self.xy_to_r`) is split into two different encoders:
  * One produces the global representation `r` (deterministic path)
  * The other produces a separate code `s` that parameterises latent `z`, instead of directly parameterising `z` with `r` (latent path)
2. The `Decoder` also accepts `r`, not only target location `x` and `z` (so it should be `xzr_to_y`)

I wonder what the reasoning behind the differences are, and why the NP model described in the NP paper differs from the NP model described in the ANP paper. I suspect difference 1 only matters because they introduced cross-attention in the deterministic path and not the latent path.

> The interpretation of the latent path is that `z` gives rise to correlations in the marginal distribution of the target predictions `y_T` , modelling the global structure of the stochastic process realisation, whereas the deterministic path models the fine-grained local structure.

As for difference 2, it probably doesn't matter too much?

The [official open-source implementation for Neural Processes](https://colab.research.google.com/github/deepmind/neural-processes/blob/master/attentive_neural_process.ipynb) also uses the architecture described in the ANP paper, not the NP paper.

Thanks for reading!
P.S. Do you know of any implementation of ANP for 2D functions?

**Comment by christabella** [comment]
Ah, it is partially explained here:
https://github.com/deepmind/neural-processes/issues/3

Setting the argument `use_deterministic_path` to `False` should remove difference 1. However, in the [official notebook](https://colab.research.google.com/github/deepmind/neural-processes/blob/master/attentive_neural_process.ipynb) this is anyway `True` for both NP and ANP it seems.

**Comment by hyunjik11** [comment]
Indeed the original NP paper only has experiments that uses the latent path, but the CNP paper has experiments that uses both the latent path and the deterministic path: c.f. section 4.2.3 of https://arxiv.org/abs/1807.01613 . In the ANP paper and the NP implementation, we chose to use both paths because we noticed that performance (both quantitative and qualitative) was best when using both. 

Also there is no official public implementation of 2D ANPs, but the paper's appendix should have all the necessary details needed to implement it.

Hope that helps!
